cupy/cusolver.py

```diff
@@ -531,3 +531,138 @@ def gesv(a, b):
     cupy.linalg.util._check_cusolver_dev_info_if_synchronization_allowed(
         getrs, dinfo)
     return b
+
+
+def gels(a, b):
+    """Solves over/well/under-determined linear systems.
+
+    Computes least-square solution to equation ``ax = b` by QR factorization
+    using cusolverDn<t>geqrf().
+
+    Args:
+        a (cupy.ndarray): The matrix with dimension ``(M, N)``.
+        b (cupy.ndarray): The matrix with dimension ``(M)`` or ``(M, K)``.
+
+    Returns:
+        cupy.ndarray:
+            The matrix with dimension ``(N)`` or ``(N, K)``.
+    """
+    if a.ndim != 2:
+        raise ValueError('a.ndim must be 2 (actual: {})'.format(a.ndim))
+    if b.ndim not in (1, 2):
+        raise ValueError('b.ndim must be 1 or 2 (actual: {})'.format(b.ndim))
+    if a.shape[0] != b.shape[0]:
+        raise ValueError('shape mismatch (a: {}, b: {}).'.
+                         format(a.shape, b.shape))
+
+    dtype = numpy.promote_types(a.dtype.char, 'f')
+    if dtype == 'f':
+        t = 's'
+    elif dtype == 'd':
+        t = 'd'
+    elif dtype == 'F':
+        t = 'c'
+    elif dtype == 'D':
+        t = 'z'
+    else:
+        raise ValueError('unsupported dtype (actual:{})'.format(a.dtype))
+
+    geqrf_helper = getattr(cusolver, t + 'geqrf_bufferSize')
+    geqrf = getattr(cusolver, t + 'geqrf')
+    trsm = getattr(cublas, t + 'trsm')
+    if t in 'sd':
+        ormqr_helper = getattr(cusolver, t + 'ormqr_bufferSize')
+        ormqr = getattr(cusolver, t + 'ormqr')
+    else:
+        ormqr_helper = getattr(cusolver, t + 'unmqr_bufferSize')
+        ormqr = getattr(cusolver, t + 'unmqr')
+
+    no_trans = cublas.CUBLAS_OP_N
+    if dtype.char in 'fd':
+        trans = cublas.CUBLAS_OP_T
+    else:
+        trans = cublas.CUBLAS_OP_C
+
+    m, n = a.shape
+    mn_min = min(m, n)
+    nrhs = b.shape[1] if b.ndim == 2 else 1
+    dev_info = cupy.empty(1, dtype=numpy.int32)
+    tau = cupy.empty(mn_min, dtype=dtype)
+    cusolver_handle = device.get_cusolver_handle()
+    cublas_handle = device.get_cublas_handle()
+    a_data_ptr = a.data.ptr
+    b_data_ptr = b.data.ptr
+    a = cupy.asfortranarray(a, dtype=dtype)
+    b = cupy.asfortranarray(b, dtype=dtype)
+    if a.data.ptr == a_data_ptr:
+        a = a.copy()
+    if b.data.ptr == b_data_ptr:
+        b = b.copy()
```

Comment:
`ndarray.copy(order='C')` (the default `order` is different than that of `numpy.copy(order='K')`) cannot be used here.  How about the following?
```suggestion
    a = a.copy(order='F')
    b = b.copy(order='F')
```